lib/datadog/core/remote/worker.rb

```diff
@@ -10,20 +10,25 @@ def initialize(interval:, &block)
           @thr = nil
 
           @starting = false
-          @stopping = false
           @started = false
+          @stop_requested = false
```

Comment:
Not fond of what the naming implies here ^^', I feel it introduces incidental complexity.

Here's my train of thought.

This sequence diagram shows how I view the state machine†:

```
{ ¬started, ¬starting, ¬stopping }     # 1.
           |
           |
        #start
           |
           V
{ ¬started, starting, ¬stopping }      # 2.
           |
           |
        success
           |
           V
{ started, ¬starting, ¬stopping }      # 3.
           |
           |
         #stop
           |
           V
{ started, ¬starting, stopping }       # 4.
           |
           |
         success
           |
           V
{ ¬started, ¬starting, ¬stopping }     # == 1.
```

In that context, `stop_requested` captures a transition (IOW `#stop` has been called) instead of representing a state. In effect this leads to a representational issue: `@stop_requested` will be forever `true` even when a restartable worker has been restarted and no `#stop` has been issued to that second `#start` state.

AIUI what `stop_requested` intends to introduce is:

- a distinction between the first and last state of the diagram, turning the last one into a 5. then making 5. equivalent to 1. by a conditional on `@allow_restart`
- a distinction between 3. and 4. which became identical due to the removal of `@stopping`.

As a consequence I feel introducing a `stopped` state would better represent all possible states:

```
{ ¬started, ¬starting, ¬stopping, ¬stopped }     # 1.
           |
           |
        #start
           |
           V
{ ¬started, starting, ¬stopping, ¬stopped }      # 2.
           |
           |
        success
           |
           V
{ started, ¬starting, ¬stopping, ¬stopped }      # 3.
           |
           |
         #stop
           |
           V
{ started, ¬starting, stopping, ¬stopped }       # 4.
           |
           |
         success
           |
           V
{ ¬started, ¬starting, ¬stopping, stopped }      # 5.
           |
           |
        #start
           |
           V
{ ¬started, starting, ¬stopping, stopped }       # 6.
           |
           |
        success
           |
           V
{ started, ¬starting, ¬stopping, ¬stopped }      # == 3.
```

It then follows that:

- `@allow_restart` is only effective when `stopped` (and not effective when `¬stopped`)
- the current _implementation_ of `stop_requested` is equivalent to `stopping OR stopped`
  - ... except when the worker is `started`; in that cas I feel a stop count (or maybe better, a start count) could be useful
- the ambiguity about `stop_requested` on a restartable worker is resolved and fully captured by `stopping`

I also feel it makes it easier to implement helper methods from these fundamental state primitives (e.g synthesising `#stop_requested` if needed)

† `success` is actually a diamond with an `error` branch, elided so as not to overload the sequence diagram, and also because it happens via exceptional code flow.

Reply:
Changed `@stop_requested` to `@stopped` in 22f33301baac1da4fc555e22a12cef66c85260fc